Support booleans, arrays of complex types, fixed values, fill bytes

tuxji opened the pull request:
In order to handle more collaborators' schemas, add support for
boolean fields, empty sequences, arrays of complex types, validation
of fixed values, and padding of explicit length complex elements with
fill bytes.  Add new collaborator schemas, test cases, and data files
to daffodil-test and update existing tests to improve code coverage.

Changelog:

In ElementBase.scala, give hasFixedLengthOf explicit Boolean type to
make its type clearer (IDE didn't infer type when I once called it).

In daffodil_main.c, initialize new PState and UState fields (position
and fill_byte).

In xml_reader.c, add strtobool function to convert an XML element's
text to a boolean.  Sort strtobool, strtodnum, strtofnum, strtonum,
and strtounum functions and PRIMITIVE_* cases in xmlNumberElem into
alphabetical order.  Add PRIMITIVE_BOOLEAN case to xmlNumberElem.

In xml_writer.c, sort PRIMITIVE_* cases in xmlNumberElem into
alphabetical order.  Add PRIMITIVE_BOOLEAN case to xmlNumberElem.

In infoset.c, sort typeCode cases in walkInfosetNode into alphabetical
order.  Add PRIMITIVE_BOOLEAN case to walkInfosetNode.

In infoset.h, sort TypeCode enumerations into alphabetical order.  Add
PRIMITIVE_BOOLEAN to TypeCode enumerations.  Add "size_t position"
fields to both PState and UState structs.  Also add "char fill_byte"
field to UState struct.

In parsers.[ch], add big-endian and little-endian 8, 16, and 32-bit
parse_<endian>_bool<length> functions to parse binary booleans with
true_rep and false_rep as additional arguments.  Since DFDL spec says
true_rep value must fit in uint32_t but we need to know whether
true_rep is defined or not, pass true_rep as int64_t instead of
uint32_t, ignore true_rep if negative, and cast true_rep to uint32_t
before using it.  Swap boolean bytes for endianness after parsing and
before comparing with true_rep and false_rep.  Use same comparison
logic as in runtime1's BinaryBooleanParserBase.parse method to conform
to DFDL spec 13.10 on parsing boolean with binary representation.  Add
parse_validate_fixed function to validate number is same as fixed
value after parse.  Make all parser functions update PState's position
field.  Add parse_fill_bytes function to skip fill bytes until end
position is reached (like runtime1, does not perform strict check that
skipped bytes have same value as fill byte).  Reorder parser functions
into alphabetical order.

In unparsers.[ch], add big-endian and little-endian 8, 16, and 32-bit
unparse_<endian>_bool<length> functions to unparse binary booleans
with true_rep and false_rep as additional arguments.  Use same
comparison logic as in runtime1's BinaryBooleanUnparserBase.unparse
method to conform to DFDL spec 13.10 on unparsing boolean with binary
representation.  Swap boolean bytes for endianness before writing
their values.  Add unparse_validate_fixed function to validate
number is same as fixed value during unparse.  Make all unparser
functions update UState's position field.  Add unparse_fill_bytes
function to unparse fill bytes until end position is reached (like
runtime1, gets fill byte's value from PState).  Reorder unparser
functions into alphabetical order.

In NestedUnion.h, replace with freshly generated code.

In ex_nums.[ch], replace with newly generated code showing that
runtime2 will pass true_rep and false_rep values to all parse and
unparse boolean calls and these values will differ depending on how
schema defines binaryBooleanFalseRep and binaryBooleanTrueRep for each
element.  Also show fixed attribute checks for boolean, float, and
integer elements.

In Runtime2CodeGenerator.scala, add BinaryBoolean case to generateCode
method and extend BinaryBooleanCodeGenerator in order to call
binaryBooleanGenerateCode method.  Reorder extends/with clauses and
case expressions into alphabetical order.

In BinaryBooleanCodeGenerator.scala, add binaryBooleanGenerateCode
method to generate C code to handle 8, 16, and 32-bit
big-endian/little-endian binary boolean fields.  Run invariant checks
on binaryBooleanFalseRep and binaryBooleanTrueRep, define trueRep as
-1 if binaryBooleanTrueRep is not defined and pass both it and
falseRep to parse_<endian>_bit<bits> calls.  Also add validation of
boolean fixed values and make code more similar in preparation for
refactoring duplicated common code from Binary*CodeGenerator traits
later.

In BinaryFloatCodeGenerator.scala, add validation of floating point
fixed values and make code more similar for refactoring later.

In BinaryIntegerKnownLengthCodeGenerator.scala, generate code to check
that the corresponding C struct member has the same value as the value
of the 'fixed' attribute during parse and unparse if there is any
'fixed' attribute.  Make code more similar for refactoring later.

In CodeGeneratorState.scala, modify addImplementation to avoid
compiler warnings for empty sequences which produce empty C structs
with no members to parse or unparse.  Make addBeforeSwitchStatements
and addAfterSwitchStatements support padding complex elements to
explicit lengths with fill bytes in addition to supporting choice
groups.  Modify addSimpleTypeStatements to ignore empty string
arguments.  Modify addComplexTypeStatements to handle arrays of
complex types.  Because schema authors don't always get base types
right, add private methods getLengthInBits and getPrimType and use
them to return what types schema authors actually want if simple
elemnts' explicit lengths don't match their base types' implicit
lengths.  Call getPrimType in addSimpleTypeERD and addFieldDeclaration
to select correct ERD type codes and C types for simple elements.  Add
PrimType.Boolean case to addSimpleTypeERD and addFieldDeclaration and
sort their cases into alphabetical order.  Add stdbool.h header to
generateCodeHeader.

In ElementParseAndUnspecifiedLengthCodeGenerator.scala, need to pass
element to addAfterSwitchStatements to let it get element's explicit
length and fill byte.

In ex_nums.dfdl.xsd, test both binaryBooleanTrueRep="" and
binaryBooleanTrueRep="1".  Add array, bigEndian, and littleEndian
boolean elements to ensure all variations of 8, 16, and 32-bit boolean
elements are tested.  Also add integer & nonNegativeInteger elements
with explicit lengths to test these types of integer elements work
too.  Sort all elements in alphabetical order by their names.  Add new
elements to validate fixed values for boolean, float, and integer
numbers.

In ex_nums.tdml, add unparse_errors test and corresponding
ex_nums_unparse_errors.xml data file to verify error message when
unparsing incorrect fixed values.

In ex_nums_parse* and ex_nums_unparse*, add data for all new elements.

Add the following new collaborator schemas, test cases, and data files
to daffodil-test:
 - ISRM_green_to_orange_60000*
 - ISRM_orange_to_green_60002*
 - MPU_green_to_orange_60004*
 - MPU_orange_to_green_60006*
 - TestIsrmGreenToOrange60000.scala
 - TestIsrmOrangeToGreen60002.scala
 - TestMpuGreenToOrange60004.scala
 - TestMpuOrangeToGreen60006.scala

In aptina_settings_* and limits_*, add new binary data and unparse XML
files for new elements added to orion-command.dfdl.xsd.

In camera_state_unparse.xml, command_unparse.xml, and
video_settings_unparse.xml, append "Decl" to root element names to
accommodate renamed element names in orion-command.dfdl.xsd.

In orion-command.dfdl.xsd, update to latest version from collaborator
with new messages AptinaSettings and Limits and all root elements
renamed to end with "Decl" instead of having types end with "Type".

In orion-command.tdml, append "Decl" to all root element names, add
test cases for new messages AptinaSettings and Limits, and reorder
test cases to follow same order as orion-command.dfdl.xsd.

In TestExNums.scala, add new unparse_errors test.

In TestOrionCommand.scala, add test cases for new messages
AptinaSettings and Limits and reorder test cases to follow same order
as orion-command.tdml.

In Rat.scala, ignore the newly added binary data files.

stevedlawrence commented on project/Rat.scala:
```diff
@@ -122,6 +122,13 @@ object Rat {
+    file("daffodil-test/src/test/resources/org/apache/daffodil/runtime2/MPU_green_to_orange_60004_parse.dat"),
+    file("daffodil-test/src/test/resources/org/apache/daffodil/runtime2/MPU_orange_to_green_60006_parse_0.dat"),
+    file("daffodil-test/src/test/resources/org/apache/daffodil/runtime2/MPU_orange_to_green_60006_parse_1.dat"),
+    file("daffodil-test/src/test/resources/org/apache/daffodil/runtime2/aptina_settings_parse.dat"),
```
Just want to confirm that these files are apache licensed? The naming implies that maybe they are coming from some tool or from a website or something and not created from scratch by you? If they are generated by a tool, the license of these files may be considered as having the same license of the tool.

tuxji replied:
No worries, I created all of these binary and XML files myself.  Sometimes I put all null bytes in the binary file, parsed it using the schema, got an XML file, edited the XML file to put some mock data into each element, and unparsed the XML file to get a better binary file.  Other times I started with another XML file, made it work, and unparsed it to get the binary file.

mbeckerle commented on daffodil-runtime2/src/main/resources/c/libruntime/parsers.c:
```diff
@@ -69,44 +113,72 @@
+        // Error message would be easier to assemble and
+        // internationalize if we used an error struct with multiple
+        // fields instead of a const char string.
+        pstate->error_msg = "Parse: Value of element does not match value of "
+                            "its 'fixed' attribute";
```
This is an interesting issue. The WIP wiki page description of the Runtime2 subset of DFDL doesn't cover this issue. 

I believe we should distinguish parse errors, runtime SDE, and validation errors. It is often important to successfully parse data because it is "well formed" yet carries invalid values in it, if only so you can display the data, and point out to users that the values are invalid. This sort of forensic analysis of data invalidity is impossible if we refuse to parse invalid data at all.

mbeckerle commented on daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/CodeGeneratorState.scala:
```diff
@@ -262,6 +288,18 @@ class CodeGeneratorState {
+      val unparseStatement =
+        s"""    ustate->fill_byte = '\\$octalFillByte';
+           |    unparse_fill_bytes(end_position, ustate);""".stripMargin
+
```
We should try to learn from the Scala back-end experience here.

We tried an approach in the Scala back-end based on 
(a) things like fill byte change infrequently
(b) so we set them into the state when they change
(c) processors use them from the state
This can be made to work, but for the scala back-end out-of-order streaming stuff, requires copying this piece of state when suspended. Had this information been in the ERD/RD structure, no copying would be needed. 

Honestly I think anything that can be done via storing a value in the ERD instead is better done that way. It's simpler, not slower - you are still one indirection away from the required value. Basically, processors should be parameterized by an ERD (or other RunitmeData object, e.g. ChoiceRuntimeData), and a State. Anything that can be in the ERD should be. Only the other things should be in the state.

tuxji commented on daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/CodeGeneratorState.scala:
```diff
@@ -362,21 +414,72 @@ class CodeGeneratorState {
+    e.elementLengthInBitsEv.constValue.get
+  }
+
+  // Because schema authors don't always get types right, allows explicit lengths to override implicit lengths
```
Hmm, I'm making the assumption in runtime2 that the generated structs and the individual fields being parsed/unparsed should have the same types (`int32_t` in both places) if the schema calls for a 4-byte integer field.  If you really think someone may want different types in the generated structs than in the data, that is going to make the code generator's job more difficult (C compiler warnings or errors might be generated).  I've added a warning as you suggested.

tuxji reviewed (commented):
Responses here and will push another commit by tomorrow.

tuxji commented:
Passing diagnostics to TDMLRunner and looking for a way to make it check runtime2 TDML warnings was more tricky than I expected.  I still had to set defaultShouldDoWarningComparisonOnCrossTests temporarily true in RunnerFactory while running runtime2's ex_nums.tdml tests to verify that I could get the generator's warnings checked appropriately.  Otherwise, the generator's warnings are never checked because TDMLRunner considers runtime2's TDML tests to be cross tests.  However, we will see generator warnings when we run "daffodil generate c -s ex_nums.dfdl.xsd" now:

    [warning] Schema Definition Warning: Ignoring PrimType integer, using short
    Schema context: be_integer16 Location line 81 column 30 in file:/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/test/resources/org/apache/daffodil/runtime2/ex_nums.dfdl.xsd
    [warning] Schema Definition Warning: Ignoring PrimType nonNegativeInteger, using unsignedInt
    Schema context: be_nonNegativeInteger32 Location line 88 column 30 in file:/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/test/resources/org/apache/daffodil/runtime2/ex_nums.dfdl.xsd
    [warning] Schema Definition Warning: Ignoring PrimType integer, using long
    Schema context: le_integer64 Location line 124 column 30 in file:/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/test/resources/org/apache/daffodil/runtime2/ex_nums.dfdl.xsd
    [warning] Schema Definition Warning: Ignoring PrimType nonNegativeInteger, using unsignedByte
    Schema context: le_nonNegativeInteger8 Location line 136 column 30 in file:/home/interran/apache/daffodil-runtime2/daffodil-runtime2/src/test/resources/org/apache/daffodil/runtime2/ex_nums.dfdl.xsd

I've addressed all the comments; please review the pull request and approve it or ask for more changes.

stevedlawrence commented on daffodil-runtime2/src/main/resources/c/libruntime/parsers.c:
```diff
@@ -69,44 +113,72 @@
+void
+parse_validate_fixed(bool same, const char *element, PState *pstate)
+{
+    if (!pstate->error_msg && !same)
```
I guess my concern is that the generated code looks something like this (copied from examples/ex_num.c, which I think is generated?):

```c
    unparse_be_float(instance->be_float[0], ustate);
    unparse_be_float(instance->be_float[1], ustate);
    unparse_be_float(instance->be_float[2], ustate);
    ...
    unparse_be_int16(instance->be_int16[0], ustate);
    unparse_be_int16(instance->be_int16[1], ustate);
    unparse_be_int16(instance->be_int16[2], ustate);
```

And each of those ``unparse_*`` functions are responsible for checking the error state and doing nothing if there's an error. Which means if the first unparse function fails, then we still end up calling all the remaining functions. Granted those functions are just doing a null pointer check so I guess is pretty minimal, but it's not nothing and could add up if there are hundreds of calls.

I don't feel strongly about this so if you feel that overhead is small enough, that's fine with me.

But there are also probably some macro solutions that would make it less ugly. For example, the above could just become something like this:

```c
    unparse_be_float(instance->be_float[0], ustate); return_on_error;
    unparse_be_float(instance->be_float[1], ustate); return_on_error;
    unparse_be_float(instance->be_float[2], ustate); return_on_error;
    ...
    unparse_be_int16(instance->be_int16[0], ustate); return_on_error;
    unparse_be_int16(instance->be_int16[1], ustate); return_on_error;
    unparse_be_int16(instance->be_int16[2], ustate); return_on_error;
```
And return_on_error is just a macro that exapnds to ``if(ustate->error) return;``. It's still a bit ugly, but does avoid that overhead.

stevedlawrence commented on daffodil-tdml-processor/src/main/scala/org/apache/daffodil/tdml/processor/Runtime2TDMLDFDLProcessor.scala:
```diff
@@ -117,8 +117,12 @@ final class Runtime2TDMLDFDLProcessorFactory private(
+        val processor = new Runtime2TDMLDFDLProcessor(tempDir, executable, generator.getDiagnostics)
+        // Sadly, TDMLRunner never checks generator diagnostics in "Right" tuple below
+        // nor does it check processor diagnostics in cross tests (runtime2's TDML tests)
+        // unless you set defaultShouldDoWarningComparisonOnCrossTests true in RunnerFactory
```
I wonder if there's a good solution for this? Perhaps we need a way to allow parser/unparseTestCases to enable this without checking the RunnerFactory? Maybe an option would to add an attribute tdml:errors/warnings that say which implementation that errors/warnings are for, e.g:
```scala
<tdml:errors implementation="daffodil">
  ...
</tdml:errors>
<tdml:errors implementation="daffodil-runtime2">
...
</tdml:errors>
```
Then errors can be checked specific to different implementations.

Not worth doing as part of this, but somethign we should maybe considered as an update to the TDML runner.

stevedlawrence commented on daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/CodeGeneratorState.scala:
```diff
@@ -316,41 +354,55 @@ class CodeGeneratorState {
+    if (initStatement.nonEmpty) structs.top.initStatements += initStatement
+    if (parseStatement.nonEmpty) structs.top.parserStatements += parseStatement
+    if (unparseStatement.nonEmpty) structs.top.unparserStatements += unparseStatement
   }
```
Ah yeah, I missed that these were strings, I thought they were Seq's. You're right, the if-statements are needed to avoid the \n.

stevedlawrence commented on daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/BinaryBooleanCodeGenerator.scala:
```diff
@@ -0,0 +1,77 @@
+    val trueRep = if (e.binaryBooleanTrueRep.isDefined) e.binaryBooleanTrueRep.getULong else -1
+    val falseRep = e.binaryBooleanFalseRep
+    val arraySize = if (e.occursCountKind == OccursCountKind.Fixed) e.maxOccurs else 0
+    val fixed = e.xml.attribute("fixed")
+    val fixedValue = if (fixed.isDefined) fixed.get.text else ""
```
The suggestion isn't so much about efficiency, but more about making is so runtimes don't need to know about the underlying XML, and that different runtimes can access the same information the same way. So the DSOM abstracts all of the XML stuff away (maybe not all, but I think that's sort of a goal).

A benefit here is if we ever wanted to change DFDL to use some other language (which there has been minor suggestions to do due to some dislike of XML/XSD), then we only need to update the DSOM--the runtimes should be untouched.

Also, if we ever change what it meanted to be "fixed" and it becomes more than just "is the fixed attribute set", then we could update the DSOM and all runtimes automatically get it.

stevedlawrence reviewed (approved):
+1 looks good! Just a couple minor comments.

mbeckerle reviewed (approved):
+1  I like where this is going. You have a good capture of the things that are TBD/pending that I would otherwise be commenting on.
